pandas/core/groupby/groupby.py

```diff
@@ -2630,6 +2630,71 @@ def _reindex_output(
 
         return output.reset_index(drop=True)
 
+    def sample(
+        self,
+        n: Optional[int] = None,
+        frac: Optional[float] = None,
+        replace: bool = False,
+        random_state=None,
+    ):
+        """
+        Return a random sample of items from each group.
+
+        You can use `random_state` for reproducibility.
+
+        .. versionadded:: 1.1.0
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of items to return. Cannot be used with `frac`.
+            Default = 1 if `frac` is None.
+        frac : float, optional
+            Fraction of items to return. Cannot be used with `n`.
+        replace : bool, default False
+            Allow or disallow sampling of the same row more than once.
+        random_state : int, array-like, BitGenerator, np.random.RandomState, optional
+            If int, array-like, or BitGenerator (NumPy>=1.17), seed for
+            random number generator
+            If np.random.RandomState, use as numpy RandomState object.
+
+        Returns
+        -------
+        Series or DataFrame
+            A new object of same type as caller containing items randomly
+            sampled within each group from the caller object.
+
+        See Also
+        --------
+        DataFrame.sample: Generate random samples from a DataFrame object.
+        numpy.random.choice: Generates a random sample from a given 1-D numpy
+            array.
+        """
+        if frac is not None and frac > 1 and not replace:
```

Comment:
Would it be possible to dispatch the implementation to `DataFrame/Series.sample` so we can share implementation (and gain the ability to support weights as well).

I think the psudoish-code would be something like

```
samples = [self._selected_obj.loc[idx].sample(...) for idx in self.indicies.values()]
return self._concat_objects(self._get_group_keys(), samples)
```

Reply:
Yeah that makes more sense